- adds throttling documentation for directory entities (#8451)
microsoftgraph/microsoft-graph-docs: concepts/throttling.md

- [People and social](/graph/api/resources/social-overview?view=graph-rest-beta)
- [Drive (OneDrive)](/graph/api/resources/drive?view=graph-rest-1.0)
- [External item (Microsoft Search)](/graph/api/resources/externalitem?view=graph-rest-beta)

> - [Directory](/graph/api/resources/directory?view=graph-rest-1.0)

Reviewer: Global: Please test the links. Resource topic file names are almost always single words, not hyphenated unless there's a prefix. These doesn't look like the right file names/URLs. Also, you don't need to include the parameter for 1.0, it will point to the v1.0 topic by default.

Author: fixed and fixed. Sorry about that I had generated the links from the metadata. All are tested now.

> - [Directory Object](/graph/api/resources/directory-object?view=graph-rest-1.0)

Reviewer: Should the link text reflect the resource name, i.e. directoryObject? At a minimum, the names shouldn't be title capped because they aren't proper nouns.

Author: fixed!

- [Endpoint](/graph/api/resources/endpoint)
- [Extension property](/graph/api/resources/extensionproperty)
- [License details](/graph/api/resources/licensedetails)
- [Group](/graph/api/resources/group)
- [Activity based timeout policy](/graph/api/resources/activitybasedtimeoutpolicy)
- [Claims mapping policy](/graph/api/resources/claimsmappingpolicy)
- [Home realm discovery policy](/graph/api/resources/homerealmdiscoverypolicy)
- [Token issuance policy](/graph/api/resources/tokenissuancepolicy)
- [Token lifetime policy](/graph/api/resources/tokenlifetimepolicy)
- [Policy base](/graph/api/resources/policybase)
- [Sts policy](/graph/api/resources/stspolicy)
- [Contract](/graph/api/resources/contract)
- [Service principal](/graph/api/resources/serviceprincipal)
- [Subscribed sku](/graph/api/resources/subscribedsku)
- [OAuth2 permission grant](/graph/api/resources/oauth2permissiongrant)
- [Organization](/graph/api/resources/organization)
- [User](/graph/api/resources/user)
- [Group setting](/graph/api/resources/groupsetting)
- [Group setting template](/graph/api/resources/groupsettingtemplate)

#### Pattern

> Throttling works by adding individual cost of requests. The sum of requests costs is then compaired against pre-determined limits. The following table documents  existing limits, reaching any of the limits will results in `429 Too Many Requests` responses. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant.

Reviewer 2: maybe it is useful to explain that we use token bucket algorithm to throttle.  Also throttling is percentage based, so we throttle percentage of their requests depending on how much they exceed their limit

Author: ```suggestion
Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. The following table documents existing limits, exceeding any of the limits will results in `429 Too Many Requests` responses. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant.
```

@fundae what do you think of this suggestion?

Reviewer 3: Suggestion: Move this part of the sentence to the end of the paragraph: "The following table documents existing limits." 
Change the rest of the sentence to: "If any of the limits are exceeded, the result is a `429 Too Many Requests` response."

Author: thanks updated!

> Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. The following table documents existing limits, exceeding any of the limits will results in `429 Too Many Requests` responses. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant.

Author: ```suggestion
Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. If any of the limits are exceeded, the response will be `429 Too Many Requests`. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant. The following table documents existing limits: 
```

| Limit type | Limit value |
| ---------- | ----------- |
| application+tenant pair | TODO |
| application | TODO |

The following table documents base requests costs, any other request not listed in the following table has a base cost of 1.

> | Operation | Request Path | Base Request Cost |

Reviewer 4: This table needs to be sorted, and explicit that `me/{operation}` is the same of `/users/{my-id}/{operation}`

Author: I think sorting the table by request path would make more sense for people searching for "what is the cost for request X". For the equivalence between me and users, there's already a note after the table stating it.

Author: sorted

| --------- | ---------- | ------------------ |
| GET | `applications` | 2 |
| GET | `applications/{id}/extensionProperties` | 2 |
| GET | `contracts` | 3 |
| POST | `directoryObjects/getByIds` |  3 |
| GET | `domains/{id}/domainNameReferences` | 4 |
| POST | `getObjectsById` | 3 |
| GET | `groups/{id}/members` | 3 |
| GET | `groups/{id}/transitiveMembers` | 5 |
| POST | `isMemberOf` | 4 |
| POST | `me/checkMemberGroups` | 4 |
| POST | `me/checkMemberObjects` | 4 |
| POST | `me/getMemberGroups` | 2 |
| POST | `me/getMemberObjects` | 2 |
| GET | `me/licenseDetails` | 2 |
| GET | `me/memberOf` | 2 |
| GET | `me/ownedObjects` | 2 |
| GET | `me/transitiveMemberOf` | 2 |
| GET | `oauth2PermissionGrants` | 2 |
| GET | `oauth2PermissionGrants/{id}` | 2 |
| GET | `servicePrincipals/{id}/appRoleAssignments` | 2 |
| GET | `subscribedSkus` | 3 |
| GET | `users` | 2 |

Other factors that impact a request cost:

- Using $select decreases cost by 1
- Using $expand increases cost by 1
- Using $top with a value of less than 20 decreases cost by 1

> > [!NOTE]

Reviewer 3: Change to: >**Note:**:  A request cost can never be lower than 1.

Author: ```suggestion
> **Note:** 
```

> A request cost can never be lower than 1.

> > [!NOTE]

Reviewer 3: Change the format of the note.

Author: ```suggestion
> **Note:** 
```

> Any request cost that applies to a request path starting with `me/` also applies to equivalent requests starting with `users/{id | userPrincipalName}/`.

#### Additional headers

##### Request headers

> - **x-ms-throttle-priority**. If the header doesn't exist or set to any other value, it indicates a normal request. We recommend limiting interactive requests. Set the priority to interactive to only the requests where the user initiated and waiting for a response. The values of this header can be the following:

Reviewer 3: Might be better to have a hyphen instead of a period: **x-ms-throttle-priority** - If the...

Author: ```suggestion
- **x-ms-throttle-priority** - If the header doesn't exist or set to any other value, it indicates a normal request. We recommend limiting interactive requests. Set the priority to interactive to only the requests where the user initiated and waiting for a response. The values of this header can be the following:
```

>     1. Low - Indicates the request is low priority. Throttling this request doesn't cause user-visible failures.

Author: ```suggestion
    - Low - Indicates the request is low priority. Throttling this request doesn't cause user-visible failures.
```

>     1. Normal - Default if no value is provided. Indicates that the request is default priority.

Author: ```suggestion
    - Normal - Default if no value is provided. Indicates that the request is default priority.
```

>     1. Interactive - Indicates that the request is interactive or high priority. Throttling this request causes user-visible failures.

Author: ```suggestion
    - Interactive - Indicates that the request is interactive or high priority. Throttling this request causes user-visible failures.
```

> > [!NOTE]

Author: ```suggestion
> **Note:** 
```

> Low requests will be throttled when 100% of the limit is reached, Normal at 110%, and Interactive at 170%. Applications using too many Interactive requests will get throttled more aggressively.

##### Regular responses requests

- **x-aad-resource-unit** - Indicates the resource unit used for this request.
- **x-aad-app-warning-limit-exceeded** - Returned only when the application consumed more than 80% of its limit. The value indicates how much the application exceeds the 80% limit. This can be used by the callers to set up an alert and take action.

##### Throttled responses requests

- **x-ms-throttle-scope** - Indicates the scope of throttling and can have the following value:

>     1. Tenant_Application_All - All requests for a particular tenant for the current application.

Reviewer 3: These should be bulleted items "-", not numbered steps.

>     1. Tenant_Application_All - All requests for a particular tenant for the current application.

Author: ```suggestion
    - Tenant_Application_All - All requests for a particular tenant for the current application.
```

>     1. Tenant_Application_Write - Requests that involve create, update, or delete operations for a particular tenant for the current application.

Author: ```suggestion
    - Tenant_Application_Write - Requests that involve create, update, or delete operations for a particular tenant for the current application.
```

>     1. Partition_Application_All - All requests for all the partition that the current tenant is in for the current application.

Author: ```suggestion
    - Partition_Application_All - All requests for all the partition that the current tenant is in for the current application.
```

>     1. Partition_Application_Write - Requests that involve create, update, or delete operations for the partition that the current tenant is in for the current application.

Author: ```suggestion
    - Partition_Application_Write - Requests that involve create, update, or delete operations for the partition that the current tenant is in for the current application.
```

>     1. Tenant_All - All requests for the current tenant, regardless of the application.

Author: ```suggestion
    - Tenant_All - All requests for the current tenant, regardless of the application.
```

>     1. Tenant_Write - Requests that involve create, update, or delete operations for the current tenant, regardless of the application.

Author: ```suggestion
    - Tenant_Write - Requests that involve create, update, or delete operations for the current tenant, regardless of the application.
```

>     1. Application_All - All requests for the current application.

Author: ```suggestion
    - Application_All - All requests for the current application.
```

- **x-ms-throttle-reason** - Indicates the reason for throttling and can have the following values:

>     1. CPULimitExceeded - Throttling is because the limit for cpu allocation is exceeded.

Author: ```suggestion
    - CPULimitExceeded - Throttling is because the limit for cpu allocation is exceeded.
```

>     1. WriteLimitExceeded - Throttling is because the write limit is exceeded.

Author: ```suggestion
    - WriteLimitExceeded - Throttling is because the write limit is exceeded.
```

>     1. RULimitExceeded - Throttling is because the limit for the allocated resource unit is exceeded.

Author: ```suggestion
    - RULimitExceeded - Throttling is because the limit for the allocated resource unit is exceeded.
```

- [Token lifetime policy](/graph/api/resources/tokenlifetimepolicy)
- [Policy base](/graph/api/resources/policybase)
- [Sts policy](/graph/api/resources/stspolicy)
- [Contract](/graph/api/resources/contract)
- [Service principal](/graph/api/resources/serviceprincipal)
- [Subscribed sku](/graph/api/resources/subscribedsku)
- [OAuth2 permission grant](/graph/api/resources/oauth2permissiongrant)
- [Organization](/graph/api/resources/organization)
- [User](/graph/api/resources/user)
- [Group setting](/graph/api/resources/groupsetting)
- [Group setting template](/graph/api/resources/groupsettingtemplate)

#### Pattern

Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. If any of the limits are exceeded, the response will be `429 Too Many Requests`. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant. The following table documents existing limits:

> | Limit type | Limit value |

Author: Two limits actually: read limit and write limit

Author: ```suggestion
| Limit type | Read limit value | ReadWrite Limit Value |
```

| ---------- | ----------- |
| application+tenant pair | TODO |
| application | TODO |

The following table documents base requests costs, any other request not listed in the following table has a base cost of 1.

> | Operation | Request Path | Base Request Cost |

Author: this needs a third column to identify which limit it counts against (read, write or both)

- [Token lifetime policy](/graph/api/resources/tokenlifetimepolicy)
- [Policy base](/graph/api/resources/policybase)
- [Sts policy](/graph/api/resources/stspolicy)
- [Contract](/graph/api/resources/contract)
- [Service principal](/graph/api/resources/serviceprincipal)
- [Subscribed sku](/graph/api/resources/subscribedsku)
- [OAuth2 permission grant](/graph/api/resources/oauth2permissiongrant)
- [Organization](/graph/api/resources/organization)
- [User](/graph/api/resources/user)
- [Group setting](/graph/api/resources/groupsetting)
- [Group setting template](/graph/api/resources/groupsettingtemplate)

#### Pattern

Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. If any of the limits are exceeded, the response will be `429 Too Many Requests`. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant. The following table documents existing limits:

> | Limit type | Read limit value | ReadWrite Limit Value |

Author: TODO; Update with actual values

> | ---------- | ----------- |

Author: ```suggestion
| ---------- | ----------- | -------------- |
```

> **Note:** 
> Any request cost that applies to a request path starting with `me/` also applies to equivalent requests starting with `users/{id | userPrincipalName}/`.

#### Additional headers

##### Request headers

- **x-ms-throttle-priority** - If the header doesn't exist or set to any other value, it indicates a normal request. We recommend limiting interactive requests. Set the priority to interactive to only the requests where the user initiated and waiting for a response. The values of this header can be the following:
    - Low - Indicates the request is low priority. Throttling this request doesn't cause user-visible failures.
    - Normal - Default if no value is provided. Indicates that the request is default priority.
    - Interactive - Indicates that the request is interactive or high priority. Throttling this request causes user-visible failures.

> **Note:** 
> Low requests will be throttled when 100% of the limit is reached, Normal at 110%, and Interactive at 170%. Applications using too many Interactive requests will get throttled more aggressively.

##### Regular responses requests

- **x-ms-resource-unit** - Indicates the resource unit used for this request.

> - **x-ms-throttle-limit-usage** - Returned only when the application consumed more than 80% of its limit. The value ranges from 80 to 180 and is a percentage of the use of the limit. This can be used by the callers to set up an alert and take action.

Author: ```suggestion
- **x-ms-throttle-limit-percentage** - Returned only when the application consumed more than 80% of its limit. The value ranges from 80 to 180 and is a percentage of the use of the limit. This can be used by the callers to set up an alert and take action.
```

- **x-ms-throttle-scope** - Indicates the scope of throttling and can have the following value:
    - Tenant_Application_All - All requests for a particular tenant for the current application.
    - Tenant_Application_Write - Requests that involve create, update, or delete operations for a particular tenant for the current application.
    - Partition_Application_All - All requests for all the partition that the current tenant is in for the current application.
    - Partition_Application_Write - Requests that involve create, update, or delete operations for the partition that the current tenant is in for the current application.
    - Tenant_All - All requests for the current tenant, regardless of the application.
    - Tenant_Write - Requests that involve create, update, or delete operations for the current tenant, regardless of the application.
    - Application_All - All requests for the current application.
- **x-ms-throttle-reason** - Indicates the reason for throttling and can have the following values:
    - CPULimitExceeded - Throttling is because the limit for cpu allocation is exceeded.
    - WriteLimitExceeded - Throttling is because the write limit is exceeded.

>     - RULimitExceeded - Throttling is because the limit for the allocated resource unit is exceeded.

Author: ```suggestion
    - ResourceUnitLimitExceeded - Throttling is because the limit for the allocated resource unit is exceeded.
```

| GET | `groups/{id}/transitiveMembers` | 5 | Read |
| POST | `isMemberOf` | 4 | Read |
| POST | `me/checkMemberGroups` | 4 | Read |
| POST | `me/checkMemberObjects` | 4 | Read |
| POST | `me/getMemberGroups` | 2 | Read |
| POST | `me/getMemberObjects` | 2 | Read |
| GET | `me/licenseDetails` | 2 | Read |
| GET | `me/memberOf` | 2 | Read |
| GET | `me/ownedObjects` | 2 | Read |
| GET | `me/transitiveMemberOf` | 2 | Read |
| GET | `oauth2PermissionGrants` | 2 | Read |
| GET | `oauth2PermissionGrants/{id}` | 2 | Read |
| GET | `servicePrincipals/{id}/appRoleAssignments` | 2 | Read |
| GET | `subscribedSkus` | 3 | Read |
| GET | `users` | 2 | Read |
| GET | Any identity path not listed in the table | 1 | Read |
| POST | Any identity path not listed in the table | 1 | ReadWrite |
| PATCH | Any identity path not listed in the table | 1 | ReadWrite |
| PUT | Any identity path not listed in the table | 1 | ReadWrite |
| DELETE | Any identity path not listed in the table | 1 | ReadWrite |

Other factors that impact a request cost:

- Using $select decreases cost by 1
- Using $expand increases cost by 1
- Using $top with a value of less than 20 decreases cost by 1

> > **Note:** A request cost can never be lower than 1.

Reviewer 4: add double space at the end of this line

Author: ```suggestion
> **Note:** A request cost can never be lower than 1.  
```

- [Contract](/graph/api/resources/contract?view=graph-rest-1.0)
- [Service Principal](/graph/api/resources/service-principal?view=graph-rest-1.0)
- [Subscribed Sku](/graph/api/resources/subscribed-sku?view=graph-rest-1.0)
- [O Auth2 Permission Grant](/graph/api/resources/o-auth2-permission-grant?view=graph-rest-1.0)
- [Organization](/graph/api/resources/organization?view=graph-rest-1.0)
- [User](/graph/api/resources/user?view=graph-rest-1.0)
- [Group Setting](/graph/api/resources/group-setting?view=graph-rest-1.0)
- [Group Setting Template](/graph/api/resources/group-setting-template?view=graph-rest-1.0)

#### Pattern

Throttling works by adding individual cost of requests. The sum of requests costs is then compaired against pre-determined limits. The following table documents  existing limits, reaching any of the limits will results in `429 Too Many Requests` responses. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant.

| Limit type | Limit value |
| ---------- | ----------- |

> | application+tenant pair | TODO |

Author: this needs to be updated whenever we get the final numbers

Author: ```suggestion
| application+tenant pair | TODO | TODO |
```

> | application | TODO |

Author: ```suggestion
| application | TODO | TODO |
```

> 

Author: ```suggestion
| tenant | TODO | TODO |

```

- [Token lifetime policy](/graph/api/resources/tokenlifetimepolicy)
- [Policy base](/graph/api/resources/policybase)
- [Sts policy](/graph/api/resources/stspolicy)
- [Contract](/graph/api/resources/contract)
- [Service principal](/graph/api/resources/serviceprincipal)
- [Subscribed sku](/graph/api/resources/subscribedsku)
- [OAuth2 permission grant](/graph/api/resources/oauth2permissiongrant)
- [Organization](/graph/api/resources/organization)
- [User](/graph/api/resources/user)
- [Group setting](/graph/api/resources/groupsetting)
- [Group setting template](/graph/api/resources/groupsettingtemplate)

#### Pattern

Throttling is based on a token bucket algorithm which works by adding individual cost of requests. The sum of requests costs is then compared against pre-determined limits. Only the requests exceeding the limits will be throttled. If any of the limits are exceeded, the response will be `429 Too Many Requests`. It is possible to receive `429 Too Many Requests` responses even when the following limits are not reached in situations when the services are under an important load or based on data volume for a specific tenant. The following table documents existing limits:

> | Limit type | Resource unit quotat | Write quotat |

Reviewer 3: ```suggestion
| Limit type | Resource unit quota | Write quota |
```
